File: app/src/main/java/me/echeung/moemoekyun/adapter/SongDetailAdapter.java
package me.echeung.moemoekyun.adapter;

import android.app.Activity;
import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;

import java.util.List;

import me.echeung.moemoekyun.App;
import me.echeung.moemoekyun.R;
import me.echeung.moemoekyun.client.model.Song;
import me.echeung.moemoekyun.databinding.SongDetailsBinding;
import me.echeung.moemoekyun.util.SongActionsUtil;

public class SongDetailAdapter extends ArrayAdapter<Song> {

    private Activity activity;

    public SongDetailAdapter(@NonNull Activity activity, List<Song> songs) {
        super(activity, 0, songs);

        this.activity = activity;
    }

    @NonNull
    @Override
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        LayoutInflater inflater = LayoutInflater.from(getContext());
        SongDetailsBinding binding;

        if (convertView == null) {
            binding = DataBindingUtil.inflate(inflater, R.layout.song_details, parent, false);
            convertView = binding.getRoot();
            convertView.setTag(binding);
        } else {
            binding = (SongDetailsBinding) convertView.getTag();
        }

        Song song = getItem(position);
        binding.setSong(song);
        binding.setIsAuthenticated(App.getAuthUtil().isAuthenticated());
        binding.setIsFavorite(song.isFavorite());

        binding.requestBtn.setOnClickListener(view -> {
            SongActionsUtil.request(activity, song);
        });

        binding.favoriteBtn.setOnClickListener(view -> {
            SongActionsUtil.toggleFavorite(activity, song);

            song.setFavorite(!song.isFavorite());
            binding.setIsFavorite(song.isFavorite());
        });

        binding.getRoot().setOnLongClickListener(view -> {
            SongActionsUtil.copyToClipboard(view.getContext(), song);
            return true;
        });

        return binding.getRoot();
    }

}

File: app/src/main/java/me/echeung/moemoekyun/adapter/SongDetailAdapter.kt
package me.echeung.moemoekyun.adapter

import android.app.Activity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import androidx.databinding.DataBindingUtil
import me.echeung.moemoekyun.App
import me.echeung.moemoekyun.R
import me.echeung.moemoekyun.client.model.Song
import me.echeung.moemoekyun.databinding.SongDetailsBinding
import me.echeung.moemoekyun.util.SongActionsUtil

class SongDetailAdapter(
        private val activity: Activity,
        songs: List<Song>,
        private val songActionsUtil: SongActionsUtil
) : ArrayAdapter<Song>(activity, 0, songs) {

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val inflater = LayoutInflater.from(context)
        val binding: SongDetailsBinding
        var view = convertView

        if (view == null) {
            binding = DataBindingUtil.inflate(inflater, R.layout.song_details, parent, false)
            view = binding.root
            view.tag = binding
        } else {
            binding = view.tag as SongDetailsBinding
        }

        val song = getItem(position) ?: return binding.root

        binding.song = song
        binding.isAuthenticated = App.authTokenUtil.isAuthenticated
        binding.isFavorite = song.favorite

        binding.requestBtn.setOnClickListener { songActionsUtil.request(activity, song) }

        binding.favoriteBtn.setOnClickListener {
            songActionsUtil.toggleFavorite(activity, song)

            song.favorite = !song.favorite
            binding.isFavorite = song.favorite
        }

        binding.root.setOnLongClickListener { v ->
            songActionsUtil.copyToClipboard(v.context, song)
            true
        }

        return binding.root
    }
}
